internal/storage/image.go

```diff
@@ -445,6 +474,88 @@ func (svc *imageService) imageStatus(systemContext *types.SystemContext, unstabl
 	return &result, nil
 }
 
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, imageName RegistryImageReference, imageID StorageImageID) error {
+	policy, err := signature.DefaultPolicy(systemContext)
+	if err != nil {
+		return fmt.Errorf("get default policy: %w", err)
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+
+	defer func() {
+		if err := policyContext.Destroy(); err != nil {
+			log.Errorf(ctx, "Error destroying policy: %+v", err)
+		}
+	}()
+
+	if err := svc.checkSignature(ctx, systemContext, policyContext, imageName, imageID); err != nil {
+		return fmt.Errorf("checking signature of %q: %w", imageName, err)
+	}
+
+	log.Debugf(ctx, "Is allowed to run config image %s (policy path: %q)", imageName, systemContext.SignaturePolicyPath)
+
+	return nil
+}
+
+func (svc *imageService) checkSignature(ctx context.Context, sys *types.SystemContext, policyContext *signature.PolicyContext, imageName RegistryImageReference, imageID StorageImageID) error {
+	userIdentityRef, err := docker.NewReference(imageName.Raw())
+	if err != nil {
+		return fmt.Errorf("creating docker:// reference for %q: %w", imageName.Raw().String(), err)
+	}
+
+	storageRef, err := imageID.imageRef(svc)
+	if err != nil {
+		return fmt.Errorf("creating containers-storage: reference for %v: %w", storageRef, err)
+	}
+	log.Debugf(ctx, "Created storageRef = %q", transports.ImageName(storageRef))
+
+	storageSource, err := storageRef.NewImageSource(ctx, sys)
+	if err != nil {
+		return fmt.Errorf("creating image source for local store image: %w", err)
+	}
+	defer storageSource.Close()
+
+	unparsedToplevel := cimage.UnparsedInstance(storageSource, nil)
+	topManifest, topMIMEType, err := unparsedToplevel.Manifest(ctx)
+	if err != nil {
+		return fmt.Errorf("get top level manifest: %w", err)
+	}
+
+	unparsedInstance := unparsedToplevel
+	if manifest.MIMETypeIsMultiImage(topMIMEType) {
+		manifestList, err := manifest.ListFromBlob(topManifest, topMIMEType)
+		if err != nil {
+			return fmt.Errorf("parsing list manifest: %w", err)
+		}
+
+		instanceDigest, err := manifestList.ChooseInstance(sys)
+		if err != nil {
+			return fmt.Errorf("choosing instance: %w", err)
+		}
+
+		unparsedInstance = cimage.UnparsedInstance(storageSource, &instanceDigest)
+	}
+
+	mixedUnparsedInstance := cimage.UnparsedInstanceWithReference(unparsedInstance, userIdentityRef)
+
+	allowed, err := policyContext.IsRunningImageAllowed(ctx, mixedUnparsedInstance)
+	if err != nil {
+		var policyErr signature.PolicyRequirementError
+		if errors.As(err, &policyErr) {
+			return fmt.Errorf("%w: %w", crierrors.ErrSignatureValidationFailed, err)
+		}
+		return fmt.Errorf("verifying signatures: %w", err)
+	}
+	if !allowed {
+		return errors.New("internal inconsistency: !allowed and no error")
+	}
```

Comment:
Should this be something more severe? Like a `panic()` or `log.Fatal()`? Would this be a "this can never happen" conditions?

Reply:
updated to log.Fatalf

Reply:
Just to make sure: would this semantic be acceptable? We will now terminate CRI-O here. So want to make sure we do the right thing...

Reply:
I mean, the new semantics/behaviour with `log.Fatalf()`. We already got a linter complaint. :smile:

Reply:
OK. I saw latest update. Makes sense.

Reply:
I updated with a direct panic which the linter doesn't complain about. my understanding is defers still run in a panic scenerio

Reply:
though I'm not sure why log.Fatalf doesn't work the same way...

Reply:
@haircommander, nice!

_For someone who stumbles here:_

The `panic()` has proper housekeeping within the runtime (panic is part of the "defer", "panic" and "recover" trio that can be used for control flow; interestingly, only "defer" here is a keyword, rest is more like `println()` [not the `fmt` package one]), whereas `log.Fatalf()` just prints (at the right log level) and then calls `os.Exit()`, which then does nothing other than calls the underlying OS-specific exit implementation, I suppose.

Reply:
FWIW the only thing I care about here is that the pull does not succeed. 

WRT the question whether CRI-O wants to panic (and presumably be restarted?) or just return an error: because this supposedly can’t happen, I don’t really know how to choose which one to prefer.

Purely hypothetically, if there were a way to trigger this code path: if it were 100% deterministic and it lead to failure, the only true effect would be an attacker triggering a CRI-O restart. Possibly that’s a performance degradation?

If there were some randomized way for the evaluation to return unexpected data… on unexpected policy rejections, it doesn’t matter. On hypothetical unexpected policy successes… as long as the node continues to make policy evaluations (i.e. whether this returns an error or panics and triggers a CRi-O restart), there’s a risk.

So I suppose the one _slightly_ better alternative would be for CRI-O to take the node down and to have it _stay down_, but that’s clearly unpalatable. And we wouldn’t be writing this new mechanism for a “should never happen” error case.

Reply:
… as a point of reference, the pull itself just returns an error:
https://github.com/containers/image/blob/31d4ad14fe4da8ef3969ff67297831ab291c76f1/copy/single.go#L77-L79

In retrospect, this two-returns API has not really been useful. c/image promises API stability so we can’t just _change_ it, but adding a new function might happen in a future.

Reply:
@mtrmac, to add a bit more colour here.

[...]
> Purely hypothetically, if there were a way to trigger this code path: if it were 100% deterministic and it lead to failure, the only true effect would be an attacker triggering a CRI-O restart. Possibly that’s a performance degradation?
[...]

The way to deploy and set things up varies, but for a typical OpenShift node, CRI-O crashing repeatedly would eventually lead to systemd marking the unit as broken, and CRI-O would remain stopped. This leads to the kubelet losing its connectivity to the CRI, eventually marking the node as NotReady. Hopefully, monitoring will pick this up, and the operator will know something is wrong. So, this can be a bit of a _saving grace_, so to speak.

An excerpt from the systemd unit for CRI-O (OpenShift 4.12.29):

```
StartLimitIntervalUSec=10s
StartLimitBurst=5
```

From [man 5 systemd.unit]( https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html) (section: _StartLimitIntervalSec=interval, StartLimitBurst=burst_):

> Configure unit start rate limiting. Units which are started more than burst times within an interval time span are not permitted to start any more. Use StartLimitIntervalSec= to configure the checking interval and StartLimitBurst= to configure how many starts per interval are allowed.

Someone would have to perform a "slow loris" type attack here to keep CRI-O crashing without triggering the systemd limits, I suppose.

Reply:
Thanks, I admit I didn’t check. In that case, terminating the process is more likely to highlight the problem, and seems a bit better to me.